qwerty_ast_to_mlir/src/lower/qpu.rs

```diff
@@ -434,44 +428,90 @@ fn is_bell_basis(vecs: &[Vector]) -> bool {
-/// Returns `true` if this is the Fourier basis as defined in the QCE '25
-/// prelude using the `revolve` basis generator.
-fn is_fourier(basis: &Basis) -> bool {
-    match basis {
-        // {'0'+'1', '0'-'1'} == pm (the fourier[1] base case)
-        Basis::BasisLiteral { vecs, .. } if vecs.len() == 2 => {
-            (if let Vector::UniformVectorSuperpos { q1, q2, .. } = &vecs[0] {
-                matches!(&**q1, Vector::ZeroVector { .. })
-                    && matches!(&**q2, Vector::OneVector { .. })
-            } else {
-                false
-            }) && (if let Vector::UniformVectorSuperpos { q1, q2, .. } = &vecs[1] {
-                matches!(&**q1, Vector::ZeroVector { .. })
-                    && if let Vector::VectorTilt { q, angle_deg, .. } = &**q2 {
-                        matches!(&**q, Vector::OneVector { .. })
-                            && angles_are_approx_equal(*angle_deg, 180.0)
-                    } else {
-                        false
-                    }
-            } else {
-                false
-            })
+/// If the Vec of Bases is a vector of only basis literals {0, 1},
+/// or only {p, m}, or only {i, j}, then we want to find this!
+fn is_vec_basis_equiv_primitive(basis_elems: &Vec<Basis>) -> (bool, qwerty::PrimitiveBasis) {
+    if basis_elems.is_empty() {
+        return (false, qwerty::PrimitiveBasis::Z);
+    }
+
+    // Define the canonical primitive bases
+    let std = (
+        Vector::ZeroVector { dbg: None },
+        Vector::OneVector { dbg: None },
+    );
+
+    let p = Vector::UniformVectorSuperpos {
+        q1: Box::new(Vector::ZeroVector { dbg: None }),
+        q2: Box::new(Vector::OneVector { dbg: None }),
+        dbg: None,
+    };
+    let m = Vector::UniformVectorSuperpos {
+        q1: Box::new(Vector::ZeroVector { dbg: None }),
+        q2: Box::new(Vector::VectorTilt {
+            q: Box::new(Vector::OneVector { dbg: None }),
+            angle_deg: 180.0,
+            dbg: None,
+        }),
+        dbg: None,
+    };
+    let pm = (p, m);
+
+    let i = Vector::UniformVectorSuperpos {
+        q1: Box::new(Vector::ZeroVector { dbg: None }),
+        q2: Box::new(Vector::VectorTilt {
+            q: Box::new(Vector::OneVector { dbg: None }),
+            angle_deg: 90.0,
+            dbg: None,
+        }),
+        dbg: None,
+    };
+    let j = Vector::UniformVectorSuperpos {
+        q1: Box::new(Vector::ZeroVector { dbg: None }),
+        q2: Box::new(Vector::VectorTilt {
+            q: Box::new(Vector::OneVector { dbg: None }),
+            angle_deg: 270.0,
+            dbg: None,
+        }),
+        dbg: None,
+    };
+    let ij = (i, j);
+
+    // helper for classifying
+    let classify = |v1: &Vector, v2: &Vector| -> Option<qwerty::PrimitiveBasis> {
+        if v1.strip_dbg() == std.0 && v2.strip_dbg() == std.1 {
+            Some(qwerty::PrimitiveBasis::Z)
+        } else if v1.strip_dbg() == pm.0 && v2.strip_dbg() == pm.1 {
```

Comment:
I don't think we should be comparing floating-point values with `==`. We need to use `angles_are_approx_equal()` for floats (e.g., `90.0` and `270.0` above).

In fairness to you, I am doing the same thing in at least 3 other places in `qwerty_ast/src/typecheck.rs` (searching `qwerty_ast/src/typecheck.rs` for `strip_dbg()` will find them). Would you be willing to create an `v1.approx_equal(v2)` method on `Basis` and `Vector` that we can use instead of `v1.strip_dbg() == v2.strip_dbg()`? Then this function I'm commenting on can be updated as can the 3 instances in `qwerty_ast/src/typecheck.rs`